sklearn/cluster/tests/test_hierarchical.py

```diff
@@ -751,3 +751,13 @@ def test_dist_threshold_invalid_parameters():
         AgglomerativeClustering(n_clusters=None,
                                 distance_threshold=1,
                                 compute_full_tree=False).fit(X)
+
+
+def test_invalid_shape_precomputed_dist_matrix():
+    # Check that an error is raised when affinity='precomputed'
+    # and a non square matrix is passed (PR #16257).
+    rng = np.random.RandomState(0)
+    X = rng.rand(5, 3)
```

Comment:
We also need to try to pass a 3D matrix (this is the 2 cases of your if statement`. You can parametrize your test:

```python
@pytest.mark.parametrize("X", [rng.rand(5, 3), rng.rand(5, 5, 5)]
def test_invalid_shape_precomputed_dist_matrix(X):
```

Reply:
There is already a check of input array dimensions at the beginning of AgglomeraticeClustering.fit(), invoking check_array at line 567 of utils.validation.py. Thus I believe adding this test is useless and, plus, it won't pass in the piece of code I was changing.

Reply:
Actually I even removed the X.ndim!=2 condition, as it never occurrs at this point, as the check of array n of dimensions is done at the beginning of fit().

Reply:
I'm ok with that